File: src/main/java/net/greenjab/fixedminecraft/registry/item/map_book/MapBookState.java
package net.greenjab.fixedminecraft.registry.item.map_book;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking;
import net.greenjab.fixedminecraft.network.MapBookPlayer;
import net.greenjab.fixedminecraft.network.MapBookSyncPayload;
import net.minecraft.resources.Identifier;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.util.datafix.DataFixTypes;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.level.saveddata.SavedData;
import net.minecraft.world.level.saveddata.SavedDataType;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class MapBookState extends SavedData {
    public ArrayList<MapBookPlayer> players = new ArrayList<>();
    public ArrayList<Integer> mapIDs = new ArrayList<>();
    public MapBookPlayer marker = new MapBookPlayer();

    public static final Codec<MapBookState> CODEC = RecordCodecBuilder.create(
           instance -> instance.group(
                            Codec.INT.listOf().optionalFieldOf("mapIDs", List.of()).forGetter( mapState -> List.copyOf(mapState.mapIDs)),
                            MapBookPlayer.CODEC.listOf().optionalFieldOf("players", List.of()).forGetter( mapState -> List.copyOf(mapState.players)),
                            MapBookPlayer.CODEC.optionalFieldOf("marker").forGetter( mapState -> java.util.Optional.ofNullable(mapState.marker))
                    ).apply(instance, MapBookState::new)
    );


    public static SavedDataType<MapBookState> createStateType(Identifier mapId) {
        return new SavedDataType<>(mapId, () -> {
            throw new IllegalStateException("Should never create an empty map saved data");
        }, CODEC, DataFixTypes.SAVED_DATA_MAP_DATA);
    }

    public MapBookState(List<Integer> maps, List<MapBookPlayer> mapBookPlayers, Optional<MapBookPlayer> marker) {
        this(new ArrayList<>(maps), new ArrayList<>(mapBookPlayers), marker.orElse(new MapBookPlayer()));
    }

    void addPlayer(Player player) {
        MapBookPlayer p = new MapBookPlayer();
        p.setPlayer(player);
        players.add(p);
    }

    public MapBookState() {
    }

    public MapBookState(ArrayList<Integer> ids, ArrayList<MapBookPlayer> players, MapBookPlayer marker) {
        mapIDs.addAll(ids);
        this.players.clear();
        this.players.addAll(players);
        this.marker = marker;
        this.setDirty();
    }

    public void sendData(MinecraftServer server, int id) {
        for (MapBookPlayer player : players) {
            ServerPlayer SPE = server.getPlayerList().getPlayerByName(player.name);
            if (SPE != null) {
                MapBookSyncPayload payload = new MapBookSyncPayload(id,
                        mapIDs.stream().mapToInt(i -> i).toArray(), (ArrayList<MapBookPlayer>) players.clone(), marker);
                ServerPlayNetworking.send(SPE, payload);
            }
        }
        MapBookStateManager.INSTANCE.getMapBookState(server, id).players.clear();
    }

    public void addMapID(int id) {
        mapIDs.add(id);
        this.setDirty();
    }
    boolean removeMapID(int id) {
        boolean hasRemoved = false;
        ArrayList<Integer> temp = new ArrayList<>();
        for (int i : mapIDs) {
            if (i==id) hasRemoved = true;
            else temp.add(i);
        }
        mapIDs.clear();
        mapIDs.addAll(temp);

        this.setDirty();
        return hasRemoved;
    }

    void update() {
        ArrayList<Integer> temp = new ArrayList<>();
        for (int i : mapIDs) {
            if (!temp.contains(i)) {
                temp.add(i);
            }
        }
        mapIDs.clear();
        mapIDs.addAll(temp);

        this.setDirty();
    }

    public void setMarker(double x, double z, String dimension) {
        marker = new MapBookPlayer();
        marker.name = "MBPmarker";
        marker.x = x;
        marker.z = z;
        marker.dimension = dimension;
    }

}

File: src/main/kotlin/net/greenjab/fixedminecraft/registry/item/map_book/MapBookState.kt
package net.greenjab.fixedminecraft.registry.item.map_book

import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking
import net.greenjab.fixedminecraft.network.MapBookPlayer
import net.greenjab.fixedminecraft.network.MapBookSyncPayload
import net.greenjab.fixedminecraft.registry.registries.ItemRegistry
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.nbt.NbtCompound
import net.minecraft.registry.RegistryWrapper.WrapperLookup
import net.minecraft.server.MinecraftServer
import net.minecraft.world.PersistentState
import java.util.Arrays


class MapBookState() : PersistentState() {
    var players: ArrayList<MapBookPlayer> = ArrayList()
    val mapIDs: java.util.ArrayList<Int> = java.util.ArrayList()

    fun addPlayer(player: PlayerEntity) {
        val p = MapBookPlayer()
        p.setPlayer(player)
        players.add(p)
    }

    constructor(ids: IntArray?, players: ArrayList<MapBookPlayer>) : this() {
        mapIDs.clear()
        mapIDs.addAll(Arrays.stream(ids).boxed().toList())
        this.players.clear()
        for (player in players) {
            this.players.add(player)
        }
        this.markDirty()
    }

    fun sendData(server: MinecraftServer, id:Int) {
        for (player in players) {
            val SPE = server.playerManager.getPlayer(player.name)
            if (SPE != null) {
                var hold = false
                for (item in SPE.handItems) {
                    if (item.isOf(ItemRegistry.MAP_BOOK)) {
                        hold = true
                    }
                }
                if (hold) {
                    ServerPlayNetworking.send(SPE, MapBookSyncPayload(id,
                        mapIDs.stream().mapToInt { i: Int? -> i!! }.toArray(), players))
                }
            }
        }
        MapBookStateManager.INSTANCE.getMapBookState(server, id)?.players=ArrayList()
    }

    override fun writeNbt(nbt: NbtCompound, lookup: WrapperLookup): NbtCompound {
        if (mapIDs.isNotEmpty()) {
            nbt.putIntArray("mapIDs", this.mapIDs)
        }

        return nbt
    }

    fun fromNbt(nbt: NbtCompound): MapBookState {
        mapIDs.clear()
        val ids = nbt.getIntArray("mapIDs")
        mapIDs.addAll(Arrays.stream(ids).boxed().toList())
        return this
    }

    fun addMapID(id: Int) {
        mapIDs.add(id)
        this.markDirty()
    }


    fun update() {
        val temp: ArrayList<Int> = ArrayList()
        for (i in mapIDs) {
            if (!temp.contains(i)) {
                temp.add(i)
            }
        }
        mapIDs.clear()
        for (i in temp){
            mapIDs.add(i)
        }

        this.markDirty()
    }


}
